File: src/store.rs
use crate::*;

pub(crate) struct Store {
    pub(crate) dir: PathBuf,
}

impl Store {
    pub(crate) fn new(state_dir: Option<PathBuf>) -> Result<Self> {
        let dir = match state_dir {
            Some(path) => path,
            None => dirs::data_local_dir()
                .context("could not resolve local data directory")?
                .join(APP_DIR),
        };
        fs::create_dir_all(&dir).with_context(|| format!("creating {}", dir.display()))?;
        Ok(Self { dir })
    }

    pub(crate) fn active_path(&self) -> PathBuf {
        self.dir.join("active.json")
    }

    pub(crate) fn pending_active_path(&self) -> PathBuf {
        self.dir.join("pending-cleanup-active.json")
    }

    pub(crate) fn acceptance_ledger_path(&self) -> PathBuf {
        self.dir.join("acceptance-ledger.json")
    }

    pub(crate) fn acceptance_followup_ledger_path(&self) -> PathBuf {
        self.dir.join("acceptance-followups.json")
    }

    pub(crate) fn acceptance_followup_reports_dir(&self) -> PathBuf {
        self.dir.join("acceptance-followups")
    }

    pub(crate) fn default_acceptance_followup_report_path(&self) -> PathBuf {
        self.acceptance_followup_reports_dir()
            .join(format!("{}.md", Local::now().date_naive()))
    }

    pub(crate) fn acceptance_event_path(&self) -> PathBuf {
        self.dir.join("acceptance-events.jsonl")
    }

    pub(crate) fn reservoir_path(&self) -> PathBuf {
        self.dir.join("candidate-reservoir.json")
    }

    pub(crate) fn event_path(&self, run: &Run) -> PathBuf {
        self.dir.join(format!("{}.jsonl", run.id))
    }

    pub(crate) fn pending_event_path(&self, run: &PendingCleanupRun) -> PathBuf {
        self.dir.join(format!("pending-cleanup-{}.jsonl", run.id))
    }

    pub(crate) fn load(&self) -> Result<Run> {
        let raw = fs::read_to_string(self.active_path()).context("loading active run")?;
        let mut run: Run = serde_json::from_str(&raw).context("parsing active run")?;
        if run.max_real_sends == 0 {
            run.max_real_sends = run.target;
        }
        Ok(run)
    }

    pub(crate) fn save(&self, run: &Run) -> Result<()> {
        let path = self.active_path();
        let tmp = path.with_extension("json.tmp");
        fs::write(&tmp, serde_json::to_string_pretty(run)?)
            .with_context(|| format!("writing {}", tmp.display()))?;
        fs::rename(&tmp, &path).with_context(|| format!("replacing {}", path.display()))?;
        Ok(())
    }

    pub(crate) fn load_pending(&self) -> Result<PendingCleanupRun> {
        let raw = fs::read_to_string(self.pending_active_path())
            .context("loading active pending-cleanup run")?;
        serde_json::from_str(&raw).context("parsing active pending-cleanup run")
    }

    pub(crate) fn save_pending(&self, run: &PendingCleanupRun) -> Result<()> {
        let path = self.pending_active_path();
        let tmp = path.with_extension("json.tmp");
        fs::write(&tmp, serde_json::to_string_pretty(run)?)
            .with_context(|| format!("writing {}", tmp.display()))?;
        fs::rename(&tmp, &path).with_context(|| format!("replacing {}", path.display()))?;
        Ok(())
    }

    pub(crate) fn load_acceptance_ledger(&self) -> Result<AcceptanceLedger> {
        let path = self.acceptance_ledger_path();
        if !path.exists() {
            return Ok(AcceptanceLedger::default());
        }
        let raw =
            fs::read_to_string(&path).with_context(|| format!("loading {}", path.display()))?;
        serde_json::from_str(&raw).with_context(|| format!("parsing {}", path.display()))
    }

    pub(crate) fn save_acceptance_ledger(&self, ledger: &AcceptanceLedger) -> Result<()> {
        let path = self.acceptance_ledger_path();
        let tmp = path.with_extension("json.tmp");
        fs::write(&tmp, serde_json::to_string_pretty(ledger)?)
            .with_context(|| format!("writing {}", tmp.display()))?;
        fs::rename(&tmp, &path).with_context(|| format!("replacing {}", path.display()))?;
        Ok(())
    }

    pub(crate) fn load_acceptance_followup_ledger(&self) -> Result<AcceptanceFollowupLedger> {
        let path = self.acceptance_followup_ledger_path();
        if !path.exists() {
            return Ok(AcceptanceFollowupLedger::default());
        }
        let raw =
            fs::read_to_string(&path).with_context(|| format!("loading {}", path.display()))?;
        serde_json::from_str(&raw).with_context(|| format!("parsing {}", path.display()))
    }

    pub(crate) fn save_acceptance_followup_ledger(
        &self,
        ledger: &AcceptanceFollowupLedger,
    ) -> Result<()> {
        let path = self.acceptance_followup_ledger_path();
        let tmp = path.with_extension("json.tmp");
        fs::write(&tmp, serde_json::to_string_pretty(ledger)?)
            .with_context(|| format!("writing {}", tmp.display()))?;
        fs::rename(&tmp, &path).with_context(|| format!("replacing {}", path.display()))?;
        Ok(())
    }

    pub(crate) fn load_reservoir(&self) -> Result<CandidateReservoir> {
        let path = self.reservoir_path();
        if !path.exists() {
            return Ok(CandidateReservoir::default());
        }
        let raw =
            fs::read_to_string(&path).with_context(|| format!("loading {}", path.display()))?;
        serde_json::from_str(&raw).with_context(|| format!("parsing {}", path.display()))
    }

    pub(crate) fn save_reservoir(&self, reservoir: &CandidateReservoir) -> Result<()> {
        let path = self.reservoir_path();
        let tmp = path.with_extension("json.tmp");
        fs::write(&tmp, serde_json::to_string_pretty(reservoir)?)
            .with_context(|| format!("writing {}", tmp.display()))?;
        fs::rename(&tmp, &path).with_context(|| format!("replacing {}", path.display()))?;
        Ok(())
    }

    pub(crate) fn seed_acceptance_from_history(
        &self,
        ledger: &mut AcceptanceLedger,
    ) -> Result<AcceptanceHistorySeedSummary> {
        let mut summary = AcceptanceHistorySeedSummary::default();
        for entry in
            fs::read_dir(&self.dir).with_context(|| format!("reading {}", self.dir.display()))?
        {
            let path = entry
                .with_context(|| format!("reading entry in {}", self.dir.display()))?
                .path();
            if path.extension().and_then(|extension| extension.to_str()) != Some("jsonl") {
                continue;
            }
            let Some(stem) = path.file_stem().and_then(|stem| stem.to_str()) else {
                continue;
            };
            let Ok(run_id) = Uuid::parse_str(stem) else {
                continue;
            };
            let Some((run_date, events)) = sent_events_from_controller_log(&path, run_id)? else {
                continue;
            };
            summary.run_logs += 1;
            summary.sent_events += events.len() as u32;
            summary.seeded += ledger.upsert_from_events(run_id, run_date, &events);
        }
        Ok(summary)
    }

    pub(crate) fn append_event<T: Serialize>(
        &self,
        run: &Run,
        kind: &str,
        payload: &T,
    ) -> Result<()> {
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(self.event_path(run))
            .context("opening event log")?;
        let event = serde_json::json!({
            "at": Local::now(),
            "run_id": run.id,
            "kind": kind,
            "payload": payload,
        });
        writeln!(file, "{}", serde_json::to_string(&event)?).context("writing event log")?;
        Ok(())
    }

    pub(crate) fn append_acceptance_event<T: Serialize>(
        &self,
        kind: &str,
        payload: &T,
    ) -> Result<()> {
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(self.acceptance_event_path())
            .context("opening acceptance event log")?;
        let event = serde_json::json!({
            "at": Local::now(),
            "kind": kind,
            "payload": payload,
        });
        writeln!(file, "{}", serde_json::to_string(&event)?)
            .context("writing acceptance event log")?;
        Ok(())
    }

    pub(crate) fn append_pending_event<T: Serialize>(
        &self,
        run: &PendingCleanupRun,
        kind: &str,
        payload: &T,
    ) -> Result<()> {
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(self.pending_event_path(run))
            .context("opening pending-cleanup event log")?;
        let event = serde_json::json!({
            "at": Local::now(),
            "run_id": run.id,
            "kind": kind,
            "payload": payload,
        });
        writeln!(file, "{}", serde_json::to_string(&event)?)
            .context("writing pending-cleanup event log")?;
        Ok(())
    }
}

File: archive/legacy-go-js/internal/app/store.go
package app

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

type Store struct {
	Dir string
}

func NewStore(stateDir string) (*Store, error) {
	dir := stateDir
	if strings.TrimSpace(dir) == "" {
		resolved, err := dataLocalDir()
		if err != nil {
			return nil, err
		}
		dir = filepath.Join(resolved, appDir)
	}
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return nil, fmt.Errorf("creating %s: %w", dir, err)
	}
	return &Store{Dir: dir}, nil
}

func dataLocalDir() (string, error) {
	if runtime.GOOS == "darwin" {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("could not resolve local data directory: %w", err)
		}
		return filepath.Join(home, "Library", "Application Support"), nil
	}
	if value := os.Getenv("XDG_DATA_HOME"); strings.TrimSpace(value) != "" {
		return value, nil
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("could not resolve local data directory: %w", err)
	}
	return filepath.Join(home, ".local", "share"), nil
}

func (s Store) ActivePath() string {
	return filepath.Join(s.Dir, "active.json")
}

func (s Store) PendingActivePath() string {
	return filepath.Join(s.Dir, "pending-cleanup-active.json")
}

func (s Store) AcceptanceLedgerPath() string {
	return filepath.Join(s.Dir, "acceptance-ledger.json")
}

func (s Store) AcceptanceFollowupLedgerPath() string {
	return filepath.Join(s.Dir, "acceptance-followups.json")
}

func (s Store) AcceptanceFollowupReportsDir() string {
	return filepath.Join(s.Dir, "acceptance-followups")
}

func (s Store) DefaultAcceptanceFollowupReportPath() string {
	return filepath.Join(s.AcceptanceFollowupReportsDir(), Today().String()+".md")
}

func (s Store) AcceptanceEventPath() string {
	return filepath.Join(s.Dir, "acceptance-events.jsonl")
}

func (s Store) ReservoirPath() string {
	return filepath.Join(s.Dir, "candidate-reservoir.json")
}

func (s Store) EventPath(run Run) string {
	return filepath.Join(s.Dir, run.ID.String()+".jsonl")
}

func (s Store) PendingEventPath(run PendingCleanupRun) string {
	return filepath.Join(s.Dir, "pending-cleanup-"+run.ID.String()+".jsonl")
}

func (s Store) Load() (Run, error) {
	var run Run
	if err := readJSONFile(s.ActivePath(), &run, "loading active run", "parsing active run"); err != nil {
		return Run{}, err
	}
	run.Normalize()
	return run, nil
}

func (s Store) Save(run Run) error {
	run.Normalize()
	return writeJSONAtomic(s.ActivePath(), run)
}

func (s Store) LoadPending() (PendingCleanupRun, error) {
	var run PendingCleanupRun
	if err := readJSONFile(s.PendingActivePath(), &run, "loading active pending-cleanup run", "parsing active pending-cleanup run"); err != nil {
		return PendingCleanupRun{}, err
	}
	run.Normalize()
	return run, nil
}

func (s Store) SavePending(run PendingCleanupRun) error {
	run.Normalize()
	return writeJSONAtomic(s.PendingActivePath(), run)
}

func (s Store) LoadAcceptanceLedger() (AcceptanceLedger, error) {
	path := s.AcceptanceLedgerPath()
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return AcceptanceLedger{Invitations: []AcceptanceInvitation{}}, nil
	}
	var ledger AcceptanceLedger
	if err := readJSONFile(path, &ledger, "loading "+path, "parsing "+path); err != nil {
		return AcceptanceLedger{}, err
	}
	ledger.Normalize()
	return ledger, nil
}

func (s Store) SaveAcceptanceLedger(ledger AcceptanceLedger) error {
	ledger.Normalize()
	return writeJSONAtomic(s.AcceptanceLedgerPath(), ledger)
}

func (s Store) LoadAcceptanceFollowupLedger() (AcceptanceFollowupLedger, error) {
	path := s.AcceptanceFollowupLedgerPath()
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return AcceptanceFollowupLedger{Drafts: []AcceptanceFollowupRecord{}}, nil
	}
	var ledger AcceptanceFollowupLedger
	if err := readJSONFile(path, &ledger, "loading "+path, "parsing "+path); err != nil {
		return AcceptanceFollowupLedger{}, err
	}
	ledger.Normalize()
	return ledger, nil
}

func (s Store) SaveAcceptanceFollowupLedger(ledger AcceptanceFollowupLedger) error {
	ledger.Normalize()
	return writeJSONAtomic(s.AcceptanceFollowupLedgerPath(), ledger)
}

func (s Store) LoadReservoir() (CandidateReservoir, error) {
	path := s.ReservoirPath()
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return CandidateReservoir{Observations: []CandidateObservation{}}, nil
	}
	var reservoir CandidateReservoir
	if err := readJSONFile(path, &reservoir, "loading "+path, "parsing "+path); err != nil {
		return CandidateReservoir{}, err
	}
	if reservoir.Observations == nil {
		reservoir.Observations = []CandidateObservation{}
	}
	for i := range reservoir.Observations {
		if len(reservoir.Observations[i].VisibleState) == 0 {
			reservoir.Observations[i].VisibleState = json.RawMessage("null")
		}
		if reservoir.Observations[i].MenuLabels == nil {
			reservoir.Observations[i].MenuLabels = []string{}
		}
	}
	return reservoir, nil
}

func (s Store) SaveReservoir(reservoir CandidateReservoir) error {
	if reservoir.Observations == nil {
		reservoir.Observations = []CandidateObservation{}
	}
	return writeJSONAtomic(s.ReservoirPath(), reservoir)
}

func (s Store) SeedAcceptanceFromHistory(ledger *AcceptanceLedger) (AcceptanceHistorySeedSummary, error) {
	summary := AcceptanceHistorySeedSummary{}
	entries, err := os.ReadDir(s.Dir)
	if err != nil {
		return summary, fmt.Errorf("reading %s: %w", s.Dir, err)
	}
	for _, entry := range entries {
		if entry.IsDir() || filepath.Ext(entry.Name()) != ".jsonl" {
			continue
		}
		stem := strings.TrimSuffix(entry.Name(), filepath.Ext(entry.Name()))
		runID, err := parseUUID(stem)
		if err != nil {
			continue
		}
		path := filepath.Join(s.Dir, entry.Name())
		runDate, events, ok, err := SentEventsFromControllerLog(path, runID)
		if err != nil {
			return summary, err
		}
		if !ok {
			continue
		}
		summary.RunLogs++
		summary.SentEvents += uint32(len(events))
		summary.Seeded += ledger.UpsertFromEvents(runID, runDate, events)
	}
	return summary, nil
}

func (s Store) AppendEvent(run Run, kind string, payload any) error {
	event := map[string]any{
		"at":      time.Now(),
		"run_id":  run.ID,
		"kind":    kind,
		"payload": payload,
	}
	return appendJSONLine(s.EventPath(run), event, "opening event log", "writing event log")
}

func (s Store) AppendAcceptanceEvent(kind string, payload any) error {
	event := map[string]any{
		"at":      time.Now(),
		"kind":    kind,
		"payload": payload,
	}
	return appendJSONLine(s.AcceptanceEventPath(), event, "opening acceptance event log", "writing acceptance event log")
}

func (s Store) AppendPendingEvent(run PendingCleanupRun, kind string, payload any) error {
	event := map[string]any{
		"at":      time.Now(),
		"run_id":  run.ID,
		"kind":    kind,
		"payload": payload,
	}
	return appendJSONLine(s.PendingEventPath(run), event, "opening pending-cleanup event log", "writing pending-cleanup event log")
}

func readJSONFile(path string, target any, readContext string, parseContext string) error {
	raw, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("%s: %w", readContext, err)
	}
	if err := json.Unmarshal(raw, target); err != nil {
		return fmt.Errorf("%s: %w", parseContext, err)
	}
	return nil
}

func writeJSONAtomic(path string, value any) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return fmt.Errorf("creating %s: %w", filepath.Dir(path), err)
	}
	tmp := strings.TrimSuffix(path, filepath.Ext(path)) + filepath.Ext(path) + ".tmp"
	raw, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	if err := os.WriteFile(tmp, raw, 0o644); err != nil {
		return fmt.Errorf("writing %s: %w", tmp, err)
	}
	if err := os.Rename(tmp, path); err != nil {
		return fmt.Errorf("replacing %s: %w", path, err)
	}
	return nil
}

func appendJSONLine(path string, value any, openContext string, writeContext string) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return fmt.Errorf("creating %s: %w", filepath.Dir(path), err)
	}
	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return fmt.Errorf("%s: %w", openContext, err)
	}
	defer file.Close()
	raw, err := json.Marshal(value)
	if err != nil {
		return err
	}
	if _, err := file.Write(append(raw, '\n')); err != nil {
		return fmt.Errorf("%s: %w", writeContext, err)
	}
	return nil
}

func readJSONLines(path string, fn func(lineNumber int, raw []byte) error) error {
	file, err := os.Open(path)
	if err != nil {
		return fmt.Errorf("opening %s: %w", path, err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	lineNumber := 0
	for scanner.Scan() {
		lineNumber++
		raw := strings.TrimSpace(scanner.Text())
		if raw == "" {
			continue
		}
		if err := fn(lineNumber, []byte(raw)); err != nil {
			return err
		}
	}
	if err := scanner.Err(); err != nil {
		return fmt.Errorf("reading %s: %w", path, err)
	}
	return nil
}
